File: ds/dynamic_array.py
class DynamicArray:
    def __init__(self):
        self.size = 2
        self.arr = [None] * self.size
        self.count = 0

    def append(self, val):
        if self.count == self.size:
            self.resize()
        self.arr[self.count] = val
        self.count += 1

    def resize(self):
        self.size = 2 * self.size
        newArr = [None] * self.size

        for i in range(self.count):
            newArr[i] = self.arr[i]
        self.arr = newArr

    def pop(self):
        if self.count <= 0:
            raise BaseException('out of index')
        index = self.count - 1
        val = self.arr[index]
        self.arr[index] = None
        self.count -= 1
        return val

    def print(self):
        if self.count <= 0:
            raise BaseException('out of index')
        for i in range(self.count):
            if self.arr[i]:
                print(self.arr[i])

# test cases
def test_dynamic_array():
    da = DynamicArray()
    da.append(1)
    da.append(2)
    da.append(3)
    da.append(4)
    da.append(5)

    da.pop()

    assert len(da.arr) == 8
    assert da.count == 4
    assert da.arr[3] == 4

    print('All test cases pass')

test_dynamic_array()

File: ds/dynamic_array.js
class DynamicArray {
  constructor() {
    this.size = 2;
    this.length = 0;
    this.arr = new Array(2);
  }

  push(n) {
    if (this.size <= this.length) this.resize();
    this.arr[this.length] = n;
    this.length++;
  }

  resize() {
    this.size = 2 * this.size;
    let newArray = new Array(this.size);

    for (let i = 0; i < this.arr.length; i++) {
      newArray[i] = this.arr[i];
    }

    this.arr = newArray;
  }

  pop() {
    if (this.length <= 0) throw new Error('out of index');
    let index = this.length - 1;
    let val = this.arr[index];
    this.arr[index] = null;
    this.length -= 1;
    return val;
  }

  print() {
    if (this.length <= 0) throw new Error('out of index');
    for (let i = 0; i < this.arr.length; i++) {
      if (this.arr[i]) console.log(this.arr[i]);
    }
  }
}

function test_dynamic_array(){
    const assert = require('assert');
    
    let dynamicArray = new DynamicArray();
    dynamicArray.push(1);
    dynamicArray.push(2);
    dynamicArray.push(3);
    dynamicArray.push(4);
    dynamicArray.push(5);

    assert(dynamicArray.length === 5);
    assert(dynamicArray.size === 8);

    let val = dynamicArray.pop();

    assert(val === 5);
    assert(dynamicArray.length === 4);

    console.log('All test cases pass');
}

test_dynamic_array();